File: workspace/CustomClusterRenderer.m
//
//  CustomClusterRenderer.m
//  GoogleMapsUtils
//
//  Created by Nikolai Borovennikov on 28.02.2020.
//  Copyright © 2020 Google. All rights reserved.
//

@import GoogleMaps;

#import "CustomClusterRenderer.h"

@implementation CustomClusterRenderer
{
    BOOL isShow;
}

- (instancetype)initWithMapView:(GMSMapView *)mapView clusterIconGenerator:(id<GMUClusterIconGenerator>)iconGenerator
{
    self = [super initWithMapView:mapView clusterIconGenerator:iconGenerator];
    if (self != nil)
    {
        self.animatesClusters = YES;
    }
    return self;
}

- (void)renderClusters:(NSArray<id<GMUCluster>> *)clusters
{
    if (isShow)
    {
        [super renderClusters:clusters];
    }
    else
    {
        [self clearClusters];
    }

}

- (void)update
{
    if (isShow)
    {
        [super update];
    }
    else
    {
        [self clearClusters];
    }
}

- (void)setIsShowMarkersAndClusters:(BOOL)newValue
{
    isShow = newValue;

    if (!isShow)
    {
        [self clearClusters];
    }
}

- (void)clearClusters
{
    for (GMSMarker *marker in self.markers)
    {
        marker.opacity = 0.0;        
        marker.tappable = false;
    }
}

@end

File: src/CustomClusterRenderer.swift
//
//  CustomClusterRenderer.swift
//  yamagoo-ios-app
//
//  Created by Artyom Popov on 02.08.2018.
//  Copyright © 2018 Artyom Popov. All rights reserved.
//

import UIKit

public enum MarkerObjStatus {
    case normal
    case pin
}

class CustomClusterRenderer: GMUDefaultClusterRenderer {

    private var isShow = true
    var tag: MarkerObjStatus = .normal

    override init(mapView: GMSMapView, clusterIconGenerator iconGenerator: GMUClusterIconGenerator) {
        super.init(mapView: mapView, clusterIconGenerator: iconGenerator)
        animatesClusters = true
    }

    override func renderClusters(_ clusters: [GMUCluster]) {
        if isShow {
            super.renderClusters(clusters)
        } else {
            clearClusters()
        }
    }

    override func update() {
        if isShow {
            super.update()
        } else {
            clearClusters()
        }
    }

    func isShowMarkersAndClusters(_ isShow : Bool) {
        self.isShow = isShow

        if !isShow {
            clearClusters()
        }
    }

    private func clearClusters() {
        for marker in markers as! [GMSMarker] {
            marker.opacity = 0.0
            marker.isTappable = false
        }
    }
}
